File: UndoOperation/MainViewModel.cs
using DevExpress.Mvvm;
using EntityFrameworkIssues.Issues;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace UndoOperation {
    public class MainViewModel : ViewModelBase {
        EntityFrameworkIssues.Issues.IssuesContext _Context;
        ObservableCollection<User> _ItemsSource;

        public ObservableCollection<EntityFrameworkIssues.Issues.User> ItemsSource {
            get {
                if(_ItemsSource == null && !IsInDesignMode) {
                    _Context = new EntityFrameworkIssues.Issues.IssuesContext();
                    _ItemsSource = new ObservableCollection<User>(_Context.Users);
                }
                return _ItemsSource;
            }
        }

        UserCopyOperationsSupporter _CopyOperationsSupporter;
        public UserCopyOperationsSupporter CopyOperationsSupporter {
            get {
                if(_CopyOperationsSupporter == null) {
                    _CopyOperationsSupporter = new UserCopyOperationsSupporter();
                }
                return _CopyOperationsSupporter;
            }
        }

        [DevExpress.Mvvm.DataAnnotations.Command]
        public void ValidateRow(DevExpress.Mvvm.Xpf.RowValidationArgs args) {
            var item = (EntityFrameworkIssues.Issues.User)args.Item;
            if(args.IsNewItem)
                _Context.Users.Add(item);
            _Context.SaveChanges();
        }
        [DevExpress.Mvvm.DataAnnotations.Command]
        public void ValidateRowDeletion(DevExpress.Mvvm.Xpf.ValidateRowDeletionArgs args) {
            var item = (EntityFrameworkIssues.Issues.User)args.Items.Single();
            _Context.Users.Remove(item);
            _Context.SaveChanges();
        }
        [DevExpress.Mvvm.DataAnnotations.Command]
        public void DataSourceRefresh(DevExpress.Mvvm.Xpf.DataSourceRefreshArgs args) {
            _ItemsSource = null;
            _Context = null;
            RaisePropertyChanged(nameof(ItemsSource));
        }
    }
}

File: VB/Undo/MainViewModel.vb
Imports DevExpress.Mvvm
Imports EntityFrameworkIssues.Issues
Imports System.Collections.ObjectModel
Imports System.Linq

Namespace UndoOperation

    Public Class MainViewModel
        Inherits ViewModelBase

        Private _CopyOperationsSupporter As UserCopyOperationSupporter

        Private _Context As IssuesContext

        Private _ItemsSource As ObservableCollection(Of User)

        Public ReadOnly Property ItemsSource As ObservableCollection(Of User)
            Get
                If _ItemsSource Is Nothing AndAlso Not IsInDesignMode Then
                    _Context = New IssuesContext()
                    _ItemsSource = New ObservableCollection(Of User)(_Context.Users)
                End If

                Return _ItemsSource
            End Get
        End Property

        Public Property CopyOperationsSupporter As UserCopyOperationSupporter
            Get
                Return _CopyOperationsSupporter
            End Get

            Private Set(ByVal value As UserCopyOperationSupporter)
                _CopyOperationsSupporter = value
            End Set
        End Property

        <DataAnnotations.Command>
        Public Sub ValidateRow(ByVal args As Xpf.RowValidationArgs)
            Dim item = CType(args.Item, User)
            If args.IsNewItem Then _Context.Users.Add(item)
            _Context.SaveChanges()
        End Sub

        <DataAnnotations.Command>
        Public Sub ValidateRowDeletion(ByVal args As Xpf.ValidateRowDeletionArgs)
            Dim item = CType(args.Items.[Single](), User)
            _Context.Users.Remove(item)
            _Context.SaveChanges()
        End Sub

        <DataAnnotations.Command>
        Public Sub DataSourceRefresh(ByVal args As Xpf.DataSourceRefreshArgs)
            _ItemsSource = Nothing
            _Context = Nothing
            RaisePropertyChanged(NameOf(MainViewModel.ItemsSource))
        End Sub

        Public Sub New()
            CopyOperationsSupporter = New UserCopyOperationSupporter()
        End Sub
    End Class
End Namespace
